On type mismatch involving associated type, suggest constraint

estebank opened the pull request:
When an associated type is found when a specific type was expected, if
possible provide a structured suggestion constraining the associated
type in a bound.

```
error[E0271]: type mismatch resolving `<T as Foo>::Y == i32`
  --> $DIR/associated-types-multiple-types-one-trait.rs:13:5
   |
LL |     want_y(t);
   |     ^^^^^^ expected `i32`, found associated type
...
LL | fn want_y<T:Foo<Y=i32>>(t: &T) { }
   |                 ----- required by this bound in `want_y`
   |
   = note:         expected type `i32`
           found associated type `<T as Foo>::Y`
help: consider constraining the associated type `<T as Foo>::Y` to `i32`
   |
LL | fn have_x_want_y<T:Foo<X=u32, Y = i32>>(t: &T)
   |                             ^^^^^^^^^
```

```
error[E0308]: mismatched types
  --> $DIR/trait-with-missing-associated-type-restriction.rs:12:9
   |
LL |     qux(x.func())
   |         ^^^^^^^^ expected `usize`, found associated type
   |
   = note:         expected type `usize`
           found associated type `<impl Trait as Trait>::A`
help: consider constraining the associated type `<impl Trait as Trait>::A` to `usize`
   |
LL | fn foo(x: impl Trait<A = usize>) {
   |                     ^^^^^^^^^^
```

Fix #71035. Related to #70908.

estebank committed:
On type mismatch involving associated type, suggest constraint

When an associated type is found when a specific type was expected, if
possible provide a structured suggestion constraining the associated
type in a bound.

```
error[E0271]: type mismatch resolving `<T as Foo>::Y == i32`
  --> $DIR/associated-types-multiple-types-one-trait.rs:13:5
   |
LL |     want_y(t);
   |     ^^^^^^ expected `i32`, found associated type
...
LL | fn want_y<T:Foo<Y=i32>>(t: &T) { }
   |                 ----- required by this bound in `want_y`
   |
   = note:         expected type `i32`
           found associated type `<T as Foo>::Y`
help: consider constraining the associated type `<T as Foo>::Y` to `i32`
   |
LL | fn have_x_want_y<T:Foo<X=u32, Y = i32>>(t: &T)
   |                             ^^^^^^^^^
```

```
error[E0308]: mismatched types
  --> $DIR/trait-with-missing-associated-type-restriction.rs:12:9
   |
LL |     qux(x.func())
   |         ^^^^^^^^ expected `usize`, found associated type
   |
   = note:         expected type `usize`
           found associated type `<impl Trait as Trait>::A`
help: consider constraining the associated type `<impl Trait as Trait>::A` to `usize`
   |
LL | fn foo(x: impl Trait<A = usize>) {
   |                     ^^^^^^^^^^
```

rust-highfive commented:
r? @davidtwco

(rust_highfive has picked a reviewer for you, use r? to override)

estebank commented on src/librustc_middle/ty/error.rs:
```diff
@@ -492,4 +505,115 @@ impl Trait for X {
+                        }
+                    }
+                }
+                for param in hir_generics.params {
+                    if self.hir().opt_local_def_id(param.hir_id).map(|id| id.to_def_id())
+                        == Some(def_id)
```
I would love it if we had a better way of doing this, but I guess this works 🤷.

oli-obk replied:
if it occurs more often, we can introduce a helper for it.

oli-obk replied:
Maybe `LocalDefId` just needs to implement `PartialEq<DefId>` and vice versa?

estebank commented on src/test/ui/suggestions/trait-with-missing-associated-type-restriction-fixable.stderr:
```diff
@@ -0,0 +1,94 @@
+error[E0308]: mismatched types
+  --> $DIR/trait-with-missing-associated-type-restriction-fixable.rs:13:13
+   |
+LL |         qux(x.func())
+   |             ^^^^^^^^ expected `usize`, found associated type
+   |
+   = note:         expected type `usize`
+           found associated type `<impl Trait as Trait>::A`
+help: consider constraining the associated type `<impl Trait as Trait>::A` to `usize`
+   |
+LL |     fn foo<'a, T: Trait + 'a>(&self, _: impl Trait, x: impl Trait<A = usize>, _: T) {
+   |                                                                  ^^^^^^^^^^^
```
This was the case I wanted to tackle.

estebank commented on src/test/ui/specialization/specialization-default-types.stderr:
```diff
@@ -21,7 +21,7 @@ LL |     Example::generate(t)
    = note:       expected struct `std::boxed::Box<T>`
            found associated type `<T as Example>::Output`
-   = note: consider constraining the associated type `<T as Example>::Output` to `std::boxed::Box<T>`
+   = help: consider constraining the associated type `<T as Example>::Output` to `std::boxed::Box<T>`
```
We should eventually handle the case of `#![feature(specialization)]`, but I won't on this PR.

Centril commented on src/librustc_middle/ty/error.rs:
```diff
@@ -392,7 +392,10 @@ impl<'tcx> TyCtxt<'tcx> {
                         db.note("you might be missing a type parameter or trait bound");
                     }
-                    (ty::Param(p), _) | (_, ty::Param(p)) => {
+                    (ty::Param(p), ty::Dynamic(..))
+                    | (ty::Dynamic(..), ty::Param(p))
+                    | (ty::Param(p), ty::Opaque(..))
+                    | (ty::Opaque(..), ty::Param(p)) => {
```
```suggestion
                    (ty::Param(p), ty::Dynamic(..) | ty::Opaque(..))
                    | (ty::Dynamic(..) | ty::Opaque(..), ty::Param(p)) => {
```

estebank replied:
done

estebank committed:
When a projection is expected, suggest constraining or calling method

estebank committed:
review comment: use or patterns

estebank committed:
Point at associated types when they have a default type

Associated types with a default type in a trait can't be relied upon to
remain of that default type when in use, so literals of that type can't
be used in the trait's items. Point at the associated type and state
that information.

Reduce verbosity for associated consts of the wrong type.

estebank commented on src/test/ui/associated-types/issue-26681.stderr:
```diff
@@ -6,7 +6,7 @@ LL |     const C: <Self::Fv as Foo>::Bar = 6665;
    = note: expected associated type `<<Self as Baz>::Fv as Foo>::Bar`
                          found type `{integer}`
-   = note: consider constraining the associated type `<<Self as Baz>::Fv as Foo>::Bar` to `{integer}` or calling a method that returns `<<Self as Baz>::Fv as Foo>::Bar`
+   = help: consider constraining the associated type `<<Self as Baz>::Fv as Foo>::Bar` to `{integer}`
```
Complex projections keep the prior output.

estebank committed:
fix rebase

estebank committed:
Suggest constraint on `impl Trait` in return type

Fix #71035.

estebank committed:
Add docs

estebank commented on src/test/ui/suggestions/trait-with-missing-associated-type-restriction.stderr:
```diff
@@ -0,0 +1,103 @@
+LL | fn foo2(x: impl Trait<i32, A = usize>) {
+   |                          ^^^^^^^^^^^
+
+error[E0308]: mismatched types
+  --> $DIR/trait-with-missing-associated-type-restriction.rs:25:12
+   |
+LL |     x.funk(3);
+   |            ^ expected associated type, found integer
+   |
+   = note: expected associated type `<T as Trait<i32>>::A`
+                         found type `{integer}`
+help: a method is available that returns `<T as Trait<i32>>::A`
+  --> $DIR/trait-with-missing-associated-type-restriction.rs:8:5
+   |
+LL |     fn func(&self) -> Self::A;
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ consider calling `Trait::func`
+help: consider constraining the associated type `<T as Trait<i32>>::A` to `{integer}`
+   |
+LL | fn bar2<T: Trait<i32, A = {integer}>>(x: T) {
+   |                     ^^^^^^^^^^^^^^^
```
For this suggestion to use `i32` instead of `{integer}` we need to run `resolve_vars_if_possible`, but that's not available in `rustc_middle` :-|

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=26375&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/26375/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/26375/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-04-15T06:10:23.9390222Z ========================== Starting Command Output ===========================
2020-04-15T06:10:23.9396311Z [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/5143e1db-5ab0-4ff3-be60-f4718193e833.sh
2020-04-15T06:10:23.9396585Z 
2020-04-15T06:10:23.9400946Z ##[section]Finishing: Disable git automatic line ending conversion
2020-04-15T06:10:23.9425199Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/71108/merge to s
2020-04-15T06:10:23.9434845Z Task         : Get sources
2020-04-15T06:10:23.9435710Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-04-15T06:10:23.9436025Z Version      : 1.0.0
2020-04-15T06:10:23.9436223Z Author       : Microsoft
---
2020-04-15T06:10:24.9293265Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-04-15T06:10:24.9299057Z ##[command]git config gc.auto 0
2020-04-15T06:10:24.9303005Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-04-15T06:10:24.9306640Z ##[command]git config --get-all http.proxy
2020-04-15T06:10:24.9313818Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/71108/merge:refs/remotes/pull/71108/merge
---
2020-04-15T06:12:44.0881811Z  ---> 78ad2f4d4aca
2020-04-15T06:12:44.0882075Z Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
2020-04-15T06:12:44.0882475Z  ---> Using cache
2020-04-15T06:12:44.0882825Z  ---> 4d2dc61c4d00
2020-04-15T06:12:44.0884242Z Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            /scripts/validate-toolstate.sh
2020-04-15T06:12:44.0885601Z  ---> 776b6266a8b7
2020-04-15T06:12:44.0885833Z Successfully built 776b6266a8b7
2020-04-15T06:12:44.0886286Z Successfully tagged rust-ci:latest
2020-04-15T06:12:44.1203244Z Built container sha256:776b6266a8b7d63e2d3c2b5a784dbf521184a904fb10bf818c6b5c7e1ab74d4a
2020-04-15T06:12:44.1203244Z Built container sha256:776b6266a8b7d63e2d3c2b5a784dbf521184a904fb10bf818c6b5c7e1ab74d4a
2020-04-15T06:12:44.1219217Z Looks like docker image is the same as before, not uploading
2020-04-15T06:12:51.7224027Z [CI_JOB_NAME=mingw-check]
2020-04-15T06:12:51.7460481Z [CI_JOB_NAME=mingw-check]
2020-04-15T06:12:51.7509646Z == clock drift check ==
2020-04-15T06:12:51.7514737Z   local time: Wed Apr 15 06:12:51 UTC 2020
2020-04-15T06:12:52.0807487Z   network time: Wed, 15 Apr 2020 06:12:52 GMT
2020-04-15T06:12:52.0829178Z Starting sccache server...
2020-04-15T06:12:52.2067478Z configure: processing command line
2020-04-15T06:12:52.2068539Z configure: 
2020-04-15T06:12:52.2069818Z configure: rust.parallel-compiler := True
---
2020-04-15T06:16:49.2314505Z     Checking rustc_span v0.0.0 (/checkout/src/librustc_span)
2020-04-15T06:16:54.1911726Z     Checking rustc_errors v0.0.0 (/checkout/src/librustc_errors)
2020-04-15T06:16:55.4817867Z     Checking rustc_feature v0.0.0 (/checkout/src/librustc_feature)
2020-04-15T06:16:55.6262855Z     Checking fmt_macros v0.0.0 (/checkout/src/libfmt_macros)
2020-04-15T06:16:55.8520953Z     Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
2020-04-15T06:16:56.5997911Z     Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
2020-04-15T06:16:56.8387618Z     Checking rustc_session v0.0.0 (/checkout/src/librustc_session)
2020-04-15T06:16:58.4341044Z     Checking rustc_attr v0.0.0 (/checkout/src/librustc_attr)
2020-04-15T06:16:58.9348625Z     Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
---
2020-04-15T06:19:00.3290144Z configure: build.submodules     := False
2020-04-15T06:19:00.3291123Z configure: rust.dist-src        := False
2020-04-15T06:19:00.3291848Z configure: build.locked-deps    := True
2020-04-15T06:19:00.3292303Z configure: llvm.assertions      := True
2020-04-15T06:19:00.3293184Z configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
2020-04-15T06:19:00.3294052Z configure: writing `config.toml` in current directory
2020-04-15T06:19:00.3294432Z configure: 
2020-04-15T06:19:00.3295034Z configure: run `python /checkout/x.py --help`
2020-04-15T06:19:00.3295424Z configure: 
---
2020-04-15T06:20:44.7185567Z Hugepagesize:       2048 kB
2020-04-15T06:20:44.7185772Z DirectMap4k:      131008 kB
2020-04-15T06:20:44.7185993Z DirectMap2M:     4063232 kB
2020-04-15T06:20:44.7186195Z DirectMap1G:     5242880 kB
2020-04-15T06:20:44.7234173Z + python3 ../x.py test src/tools/expand-yaml-anchors
2020-04-15T06:20:46.1853761Z Ensuring the YAML anchors in the GitHub Actions config were expanded
2020-04-15T06:20:46.1853761Z Ensuring the YAML anchors in the GitHub Actions config were expanded
2020-04-15T06:20:46.1861774Z Building stage0 tool expand-yaml-anchors (x86_64-unknown-linux-gnu)
2020-04-15T06:20:46.4257539Z    Compiling unicode-xid v0.2.0
2020-04-15T06:20:46.5499692Z    Compiling syn v1.0.11
2020-04-15T06:20:47.4961681Z    Compiling linked-hash-map v0.5.2
2020-04-15T06:20:47.5258676Z    Compiling lazy_static v1.4.0
2020-04-15T06:20:47.5258676Z    Compiling lazy_static v1.4.0
2020-04-15T06:20:47.7332086Z    Compiling yaml-rust v0.4.3
2020-04-15T06:20:52.4426769Z    Compiling quote v1.0.2
2020-04-15T06:21:08.0926445Z    Compiling thiserror-impl v1.0.5
2020-04-15T06:21:13.2754903Z    Compiling thiserror v1.0.5
2020-04-15T06:21:13.3338584Z    Compiling yaml-merge-keys v0.4.0
2020-04-15T06:21:15.0981595Z    Compiling expand-yaml-anchors v0.1.0 (/checkout/src/tools/expand-yaml-anchors)
2020-04-15T06:21:17.8629293Z Build completed successfully in 0:00:33
2020-04-15T06:21:17.8726315Z + python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu
2020-04-15T06:21:18.1574516Z     Finished dev [unoptimized] target(s) in 0.19s
2020-04-15T06:21:19.3128584Z Checking rustdoc artifacts (x86_64-unknown-linux-gnu -> i686-pc-windows-gnu)
---
2020-04-15T06:23:35.3620352Z     Checking rustc_feature v0.0.0 (/checkout/src/librustc_feature)
2020-04-15T06:23:35.3911196Z     Checking fmt_macros v0.0.0 (/checkout/src/libfmt_macros)
2020-04-15T06:23:35.5978535Z     Checking rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
2020-04-15T06:23:35.8087954Z     Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
2020-04-15T06:23:36.2753741Z     Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
2020-04-15T06:23:38.7435985Z     Checking rustc_attr v0.0.0 (/checkout/src/librustc_attr)
2020-04-15T06:23:39.2539768Z     Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
2020-04-15T06:23:41.4218195Z     Checking rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
2020-04-15T06:23:41.8906171Z     Checking rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
2020-04-15T06:28:14.2239526Z Done!
2020-04-15T06:28:14.2239697Z some tidy checks failed
2020-04-15T06:28:14.2239865Z 
2020-04-15T06:28:14.2239968Z 
2020-04-15T06:28:14.2241270Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo"
2020-04-15T06:28:14.2242033Z 
2020-04-15T06:28:14.2242135Z 
2020-04-15T06:28:14.2251036Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
2020-04-15T06:28:14.2251455Z Build completed unsuccessfully in 0:00:39
2020-04-15T06:28:14.2251455Z Build completed unsuccessfully in 0:00:39
2020-04-15T06:28:14.2358307Z == clock drift check ==
2020-04-15T06:28:14.2374086Z   local time: Wed Apr 15 06:28:14 UTC 2020
2020-04-15T06:28:14.5434236Z   network time: Wed, 15 Apr 2020 06:28:14 GMT
2020-04-15T06:28:16.0315095Z 
2020-04-15T06:28:16.0315095Z 
2020-04-15T06:28:16.0373338Z ##[error]Bash exited with code '1'.
2020-04-15T06:28:16.0391913Z ##[section]Finishing: Run build
2020-04-15T06:28:16.0441622Z ##[section]Starting: Checkout rust-lang/rust@refs/pull/71108/merge to s
2020-04-15T06:28:16.0447322Z Task         : Get sources
2020-04-15T06:28:16.0447695Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
2020-04-15T06:28:16.0448042Z Version      : 1.0.0
2020-04-15T06:28:16.0448279Z Author       : Microsoft
2020-04-15T06:28:16.0448279Z Author       : Microsoft
2020-04-15T06:28:16.0448647Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
2020-04-15T06:28:16.0449080Z ==============================================================================
2020-04-15T06:28:16.4202617Z Cleaning any cached credential from repository: rust-lang/rust (GitHub)
2020-04-15T06:28:16.4247276Z ##[section]Finishing: Checkout rust-lang/rust@refs/pull/71108/merge to s
2020-04-15T06:28:16.4349824Z Cleaning up task key
2020-04-15T06:28:16.4351036Z Start cleaning up orphan processes.
2020-04-15T06:28:16.4568733Z Terminate orphan process: pid (3866) (python)
2020-04-15T06:28:16.4845557Z ##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

estebank committed:
trailing newlines

estebank commented:
r? @nikomatsakis

bors commented:
:umbrella: The latest upstream changes (presumably #69793) made this pull request unmergeable. Please resolve the merge conflicts.

Dylan-DPC commented:
r? @oli-obk

bors commented:
:umbrella: The latest upstream changes (presumably #69274) made this pull request unmergeable. Please resolve the merge conflicts.

rust-highfive commented:
The job `mingw-check` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=28135&view=logs&jobId=74054d28-c774-5fab-ecc2-352fe71b230e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/28135/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/28135/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/3e6b66a3-9d80-4cb0-962a-41c16f50c6ab.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/71108/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/71108/merge:refs/remotes/pull/71108/merge
---
 ---> f7353ccad5b1
Step 6/7 : ENV RUN_CHECK_WITH_PARALLEL_QUERIES 1
 ---> Using cache
 ---> ed38efbaa060
Step 7/7 : ENV SCRIPT python3 ../x.py test src/tools/expand-yaml-anchors &&            python3 ../x.py check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu &&            python3 ../x.py build --stage 0 src/tools/build-manifest &&            python3 ../x.py test --stage 0 src/tools/compiletest &&            python3 ../x.py test src/tools/tidy &&            /scripts/validate-toolstate.sh
 ---> c5008ef7ae8e
Successfully built c5008ef7ae8e
Successfully tagged rust-ci:latest
Built container sha256:c5008ef7ae8e94d7ef502e3cef26e61208e14ebdb36913f3a8bb86291bd6430b
---
    Checking rustc_feature v0.0.0 (/checkout/src/librustc_feature)
    Checking fmt_macros v0.0.0 (/checkout/src/libfmt_macros)
    Checking rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
    Checking rustc_hir v0.0.0 (/checkout/src/librustc_hir)
    Checking rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
    Checking rustc_attr v0.0.0 (/checkout/src/librustc_attr)
    Checking rustc_parse v0.0.0 (/checkout/src/librustc_parse)
    Checking rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
    Checking rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
    Checking rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
    Checking rustc_ast_passes v0.0.0 (/checkout/src/librustc_ast_passes)
    Checking rustc_expand v0.0.0 (/checkout/src/librustc_expand)
    Checking rustc_builtin_macros v0.0.0 (/checkout/src/librustc_builtin_macros)
error[E0308]: mismatched types
   --> src/librustc_middle/ty/error.rs:701:68
    |
701 |         if let (Some(hir_id), false) = (self.hir().as_local_hir_id(body_owner_def_id), suggested) {

error[E0308]: mismatched types
error[E0308]: mismatched types
   --> src/librustc_middle/ty/error.rs:701:17
    |
701 |         if let (Some(hir_id), false) = (self.hir().as_local_hir_id(body_owner_def_id), suggested) {
    |                 ^^^^^^^^^^^^           ---------------------------------------------------------- this expression has type `(rustc_hir::hir_id::HirId, bool)`
    |                 expected struct `rustc_hir::hir_id::HirId`, found enum `std::option::Option`
    |
    = note: expected struct `rustc_hir::hir_id::HirId`
                 found enum `std::option::Option<_>`
                 found enum `std::option::Option<_>`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_middle`.

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--color" "always" "--features" " llvm" "--manifest-path" "/checkout/src/rustc/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:04:20
== clock drift check ==
  local time: Sun May  3 01:32:46 UTC 2020
  local time: Sun May  3 01:32:46 UTC 2020
  network time: Sun, 03 May 2020 01:32:46 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/71108/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/71108/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (4494) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

oli-obk commented on src/librustc_middle/ty/error.rs:
```diff
@@ -513,4 +513,344 @@ impl Trait for X {
+                _ => {}
+            }
+        }
+        if let ty::Opaque(def_id, _) = proj_ty.self_ty().kind {
```
why is this done without a condition on `suggested`?

estebank replied:
The logic for the suggestions was already mutually exclusive. Now with the change to early returns this shouldn't be ambiguous anymore.

oli-obk commented on src/librustc_middle/ty/error.rs:
```diff
@@ -513,4 +513,344 @@ impl Trait for X {
+            "consider constraining the associated type `{}` to `{}`",
+            values.expected, values.found
+        );
+        let mut suggested = false;
```
modulo my comment about why opaque types don't check `suggested`, this flag is unnecessary and the function could just use early returns.

oli-obk commented on src/librustc_middle/ty/error.rs:
```diff
@@ -513,4 +513,344 @@ impl Trait for X {
+                    kind: hir::ImplItemKind::Const(..) | hir::ImplItemKind::TyAlias(..),
+                    ..
+                }),
+            ) => false,
```
I'm not too happy about the above blacklist, it seems like we could easily regress this diagnostic without realizing it. A whitelist would be better I think.

oli-obk commented on src/librustc_middle/ty/error.rs:
```diff
@@ -513,4 +513,344 @@ impl Trait for X {
+            // We do not want to suggest calling functions when the reason of the
+            // type error is a comparison of an `impl` with its `trait` or when the
+            // scope is outside of a `Body`.
+        } else {
```
this block and the other big `if` statement later in this function could become their own functions, making it easier to read this function.

oli-obk commented on src/librustc_middle/ty/error.rs:
```diff
@@ -513,4 +513,344 @@ impl Trait for X {
+        msg: &str,
+    ) -> bool {
+        // FIXME: we would want to call `resolve_vars_if_possible` on `ty` before suggesting.
+        for bound in bounds {
+            match bound {
+                hir::GenericBound::Trait(ptr, hir::TraitBoundModifier::None) => {
+                    // Relate the type param against `T` in `<A as T>::Foo`.
+                    if ptr.trait_ref.trait_def_id() == Some(trait_ref.def_id)
+                        && self.constrain_associated_type_structured_suggestion(
+                            db, ptr.span, assoc, ty, msg,
+                        )
+                    {
+                        return true;
+                    }
+                }
+                _ => {}
+            }
+        }
+        false
```
```suggestion
		bounds.iter().any(|bound| match bound {
                hir::GenericBound::Trait(ptr, hir::TraitBoundModifier::None) => {
                    // Relate the type param against `T` in `<A as T>::Foo`.
                    ptr.trait_ref.trait_def_id() == Some(trait_ref.def_id)
                        && self.constrain_associated_type_structured_suggestion(
                            db, ptr.span, assoc, ty, msg,
                        )
                }
                _ => false,
            })
```

oli-obk commented on src/test/ui/issues/issue-69306.stderr:
```diff
@@ -8,8 +8,6 @@ LL |     const C: S0<u8> = Self(0);
    |
    = note: expected type parameter `T`
                         found type `{integer}`
-   = help: type parameters must be constrained to match other types
```
maybe we should keep a `help` around that informs the user that a type parameter can never be equal to a concrete type, even if the caller would always choose that specific type?

estebank replied:
I would prefer for us to detect these kind of cases proactively and give a more targeted error. In the case of

```rust
struct S0<T>(T);
impl<T> S0<T> {
    const C: S0<u8> = Self(0);

    fn foo() {
        Self(0);
    }
}
```

we should come up with a more extensive explanation of what `Self` resolves to, something closer to

```
error[E0308]: mismatched types
  --> $DIR/issue-69306.rs:5:28
   |
LL | impl<T> S0<T> {
   |      - this type parameter
LL |     const C: S0<u8> = Self(0);
   |                       ---- ^ expected type parameter `T`, found integer
   |                       |
   |                       expected `T` because of this
   |
   = note: expected type parameter `T`
                        found type `{integer}`
note: `Self` resolves to `S0<T>`, which doesn't match `S0<{integer}>` becaue it is more general
   |
LL |     const C: S0<u8> = Self(0);
   |                       ^^^^

error[E0308]: mismatched types
  --> $DIR/issue-69306.rs:5:28
   |
LL | impl<T> S0<T> {
   |      - this type parameter
LL |     const C: S0<u8> = Self(0);
   |              ------   ^^^^^^^ expected `u8`, found type parameter `T`
   |              |
   |              expected `u8` because of this
   |
   = note: expected type parameter `T`
                        found type `{integer}`
note: `Self` resolves to `S0<T>`, which doesn't match `S0<u8>` becaue it is more general
   |
LL |     const C: S0<u8> = Self(0);
   |                       ^^^^

error[E0308]: mismatched types
  --> $DIR/issue-69306.rs:10:14
   |
LL | impl<T> S0<T> {
   |      - this type parameter
...
LL |         Self(0);
   |              ^ expected type parameter `T`, found integer
   |
   = note: expected type parameter `T`
                        found type `{integer}`
```

estebank committed:
fix rebase

estebank committed:
review comments: move logic to their own methods

estebank committed:
review comment: use early return

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit b368229d9bbaa840e777d33e36e649967e7ecb04 has been approved by `oli-obk`

<!-- @bors r=oli-obk b368229d9bbaa840e777d33e36e649967e7ecb04 -->
<!-- homu: {"type":"Approved","sha":"b368229d9bbaa840e777d33e36e649967e7ecb04","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit b368229d9bbaa840e777d33e36e649967e7ecb04 with merge d6823ba1666fa5f65e5fdd17cfc78ff227c092f2...
<!-- homu: {"type":"BuildStarted","head_sha":"b368229d9bbaa840e777d33e36e649967e7ecb04","merge_sha":"d6823ba1666fa5f65e5fdd17cfc78ff227c092f2"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=28264)
Approved by: oli-obk
Pushing d6823ba1666fa5f65e5fdd17cfc78ff227c092f2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=28264"},"merge_sha":"d6823ba1666fa5f65e5fdd17cfc78ff227c092f2"} -->
